### PR TITLE
Implements data transformations in TSA module

### PR DESCRIPTION
**Pull request by j-bryan**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
#2120 

##### What are the significant changes in functionality due to this change request?
A new "Transformer" class is added to the TSA module which can apply any transformer class whose interface is defined by the sklearn.base.TransformerMixin base class to any target or group of target variables. The specifics of how the residual and composite signals are assembled are changed slightly to allow for non-additive signal composition so that nonlinear transformation functions can be accommodated.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by j-bryan** [comment]
@dylanjm and @GabrielSoto-INL , I'd love to get your input on how this is implemented.

File: doc/user_manual/generated/internalRom.tex [hunk 1]
```diff
@@ -2175,31 +2182,134 @@ \subsubsection{SyntheticHistory}
...
+        \begin{itemize}
+          \item \xmlAttr{target}: \xmlDesc{comma-separated strings, required}, 
+            indicates the variables for which this algorithm will be used for characterization.
+          \item \xmlAttr{seed}: \xmlDesc{integer, optional}, 
+            sets a seed for the underlying random number generator, if present.
+          \item \xmlAttr{nQuantiles}: \xmlDesc{integer, optional}, 
+            number of quantiles to use in the transformation. If \xmlAttr{nQuantiles}
+            is greater than the number of data, then the number of data is used instead. \default{1000}
+          \item \xmlAttr{outputDistribution}: \xmlDesc{[normal, uniform], optional}, 
+            distribution to transform to. Must be either 'normal' or 'uniform'. \default{normal}
```
Comment (PaulTalbot-INL) [inline doc/user_manual/generated/internalRom.tex:2312]: did you check if this compiles as expected?

File: tests/framework/unit_tests/TSA/testFunctionTransformers.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    @ In, printComment, bool, optional, if False then don't print comment
+  """
+  if res:
+    results['pass'] += 1
+  else:
+    results['fail'] += 1
+    if printComment:
+      print(f'checking {str(dtype)} {comment} | {value} != {expected}')
+
+def checkFloat(comment, value, expected, tol=1e-10, update=True):
```
Comment (PaulTalbot-INL) [inline tests/framework/unit_tests/TSA/testFunctionTransformers.py:56]: FWIW, I've really wanted to move to using pytest and standardize all these checks across RAVEN unit tests, but have never had the bandwidth to do it. It's not required (by any means) for this work, but if there's spare bandwidth and interest, it would improve the unit tests dramatically.

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
A few minor initial changes to consider; I haven't carefully checked all the implementation, but I did notice a few things.

I'm still not convinced `fit` increases readability over `characterize`, but I'm not completely against it. I do have questions about what it means to be a `generator`, however.

File: ravenframework/TSA/Transformers/FunctionTransformers.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                               same order as self.target
+      @ In, params, dict, training parameters as from self.characterize
+      @ In, pivot, np.array, time-like array values
+      @ In, settings, dict, additional settings specific to algorithm
+      @ Out, residual, np.array, reduced signal shaped [pivotValues, targets]
+    """
+    # Check for non-positive values in targets before handing off to super
+    for t, (target, data) in enumerate(params.items()):
+      if np.any(initial[:, t] <= 0):
+        raise ValueError('Log transformation requires strictly positive values!')
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/Transformers/FunctionTransformers.py:60]: That's why I also implemented the arcsinh transformation. It's still a logarithmic function but its domain includes negative values. I've seen arcsinh used for "log" transformations of electricity price data in the literature specifically to allow for the possibility of negative prices.

File: ravenframework/TSA/TimeSeriesAnalyzer.py [hunk 4]
```diff
@@ -101,23 +118,86 @@ def setDefaults(self, settings):
...
+      @ In, params, dict, parameters from training this ROM
+      @ Out, None
+    """
+
+
+class TimeSeriesGenerator(TimeSeriesAnalyzer):
+  """
+    Act as a mix-in class for algorithms that can generate synthetic time histories. This is
+    reserved exclusively for stochastic algorithms. Deterministic generative algorithms should NOT
+    inherit from this class.
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/TimeSeriesAnalyzer.py:146]: Several other comments you've made revolve around this distinction of generators needing to be stochastic or not, so I'll respond to that general idea here. My thinking there was that it can be helpful to know if an algorithm is stochastic or deterministic, and having that be difference of type was a handy way to make that distinction. I also noticed that deterministic functions tend to have a residual from fitting that we want to pass on to downstream algorithms, while generators (thinking about ARMA and GANs) often don't have that. Having stochastic and determinstic algorithms have different methods (from different base classes) would make sense in that case. In the end, the implementations for training and evaluation in TSAUser don't require there to be a stochastic generator in the list of algorithms, so maybe there's no concrete benefit from the distinction right now.

File: tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/LogARMA_0.csv
[the remark this answers is not in the document]
Comment (j-bryan) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/LogARMA_0.csv:1]: I was just following the pattern set with the ARMA_A_0.csv and FourierARMA_A_0.csv files, but the number of timesteps in all of those tests could probably be reduced without hurting the quality of the test. I'll look into it.

File: doc/user_manual/generated/internalRom.tex (hunk 1 above)
Comment (j-bryan) [reply doc/user_manual/generated/internalRom.tex:2312]: I don't have a full regression test that looks at that, but in the testNormalization.py unit tests included in the PR, both options are tested. If you're talking about the actual docstring loading properly into the user manual, it seems to compile as expected.
image

Comment (j-bryan) [reply doc/user_manual/generated/internalRom.tex:2312]: Though the single quotes are a little wonky, I suppose.

File: ravenframework/SupervisedLearning/SyntheticHistory.py [hunk 5]
```diff
@@ -78,6 +78,8 @@ def _handleInput(self, paramInput):
     """
     SupervisedLearning._handleInput(self, paramInput)
     self.readTSAInput(paramInput)
+    # if not self.canGenerate():
+    #   self.raiseAnError(IOError, 'SyntheticHistory ROM must contain at least one generative TSA algorithm!')
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/SupervisedLearning/SyntheticHistory.py:82]: Sorry, kind of an idea that I can commented out and forgot to remove. I had talked over the idea of requiring a SyntheticHistory ROM to contain a generator (stochastic or not, depending on the convention we end up settling on there), while performing a similar check to see if a TSACharacterizer contains at least one characterizing algorithm. I'll either remove the commented check or reinstate it, depending on the generator vs. transformer convention we end up following.

File: ravenframework/TSA/Fourier.py [hunk 6]
```diff
@@ -192,24 +246,6 @@ def getParamsAsVars(self, params):
           rlz[f'{baseWave}__{stat}'] = value
     return rlz
 
-
-  def generate(self, params, pivot, settings):
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/Fourier.py:196]: With Fourier being implemented as a TimeSeriesTransformer instead of a TimeSeriesGenerator, I wanted to avoid using the same `generate()` method name used in TimeSeriesGenerator. If we revert this back to being a generator (and relaxing the restriction on generators being stochastic), I'll revert the name change.

File: ravenframework/TSA/Transformers/Normalizers.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  """ Wrapper of sklearn.preprocessing.MaxAbsScaler """
+  _features = ['scale']
+  templateTransformer = skl.MaxAbsScaler()
+
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for
+      class cls.
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/Transformers/Normalizers.py:38]: Just to check the RAVEN format here, does every docstring require both an @In and an @Out even if there is no input or return value? FWIW the quoted docstring was a direct copy/paste of the matching docstring in ARMA.py.

File: ravenframework/TSA/Transformers/Filters.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  """ Base class for transformers which filter or mask data """
+
+  def __init__(self):
+    """ Constructor  """
+    super().__init__()
+
+  def handleInput(self, spec):
+    """
+      Reads user inputs into this object.
+      @ In, inp, InputData.InputParams, input specifications
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/Transformers/Filters.py:36]: Like another comment, this was a copy/paste from ARMA.py with an issue I didn't catch. I'll fix it there, too.

File: doc/user_manual/generated/internalRom.tex (hunk 1 above)
Comment (PaulTalbot-INL) [reply doc/user_manual/generated/internalRom.tex:2312]: LGTM.

File: ravenframework/TSA/Fourier.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/TSA/Fourier.py:196]: Okay, help me understand that, and maybe talking this over in a meeting makes more sense. Fourier projects the signal from signal space to a Hilbert space, but the result is not a transformation of the data, but a characterization of the data, as it seems to me. Help me see how you're seeing Fourier as a transformer.

File: ravenframework/TSA/TimeSeriesAnalyzer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/TSA/TimeSeriesAnalyzer.py:146]: I can see that it would be beneficial to have stochastic generators distinct. I would rather see `Generators` further subdivided into `Deterministic` and `Stochastic` if necessary, rather than make all deterministic `Generators` into `Transformers`. This is related to the discussion on where Fourier gets classified, above.

As to residuals, we actually do see a future where stochastic signals can subtract out the parts of the signal that are "likely" attributed to the noise characteristics, and residual pieces passed downstream. We haven't had a case to use this on yet, but it seems like it would be necessary to complete the toolset. I don't know how we'd go about a GAN informing us the likelihood a sample originated from its sampling, but for ARIMA it should be straightforward, I think.

File: ravenframework/TSA/Transformers/FunctionTransformers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/TSA/Transformers/FunctionTransformers.py:60]: Interesting, can we add a suggestion to the error to consider using arcsinh if we see negatives? That's interesting feedback for the user, who probably doesn't know about a lot of functional transforms.

File: ravenframework/TSA/Transformers/Normalizers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/TSA/Transformers/Normalizers.py:38]: Yes, I have lots of issues with our docstring militance, but it is what it is. Someone should have called me out on the ARMA one when they reviewed it. Technically, until we can get that standard overturned, the entry
```
@ In, None
```
is required if there are no inputs.

File: ravenframework/TSA/Fourier.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/Fourier.py:196]: I see it as a transformer in the sense that you give it a signal and it returns a modified signal. It certainly characterizes as well, but I see the characterization of the signal as separate thing from the operation of modifying the signal based on the results of that characterization. I'd certainly be open to discussing this and some of the broader design decisions in a meeting instead of through a Github PR review..

File: tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/LogARMA_0.csv
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/LogARMA_0.csv:1]: I think someone expanded the ARMA and Fourier tests when someone (possibly you?) pointed out how ineffective ARMA training is on small samples. If it takes a 1000-length history to be effective in testing, let's do it. In the cases it's not necessary, let's trim it down, if that makes sense.

[the remark this answers is not in the document]
Comment (j-bryan) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/LogARMA_0.csv:1]: I think histories of ~100 samples should be plenty for estimating an ARMA model. Any tests that segment the data first are going to need longer histories to make sure each segment has enough data. Looking through the tests there, the arma test and my new log_arma test shouldn't need such long histories, but we'll probably have to keep at least one long history in the training data for some of the other tests.

File: ravenframework/TSA/TimeSeriesAnalyzer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/TimeSeriesAnalyzer.py:146]: I'd be fine narrowing the scope of what is a transformer and adding some distinction for `Generators` between deterministic and stochastic generators. It seems that every algorithm would need `getResidual` and `getComposite` methods then, so we might not need the `Transformer` base class at all then, since those are the only unique methods that the class adds.

What would you think of adding those methods to the `TimeSeriesAnalyzer` base class, dropping the `Transformer` base class entirely, and making classes which are exclusively `Transformer`s right now (e.g. `LogTransformer`) direct subclasses of `TimeSeriesAnalyzer`? Then the distinction of stochastic vs deterministic generators couldbe easily handled with a class attribute/checker function within `Generators` since we probably don't want a different interface between the two.

File: ravenframework/TSA/AlgorithmClassification.png
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/TSA/AlgorithmClassification.png:1]: Updated: I see Fourier, PolyReg, and ARMA as characterizers and generators; are they also Transformers?

**Review by PaulTalbot-INL — COMMENTED** [review]
Only one discussion point, on the Venn diagram of algorithms. Thanks for making those changes.

File: ravenframework/TSA/AlgorithmClassification.png
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/AlgorithmClassification.png:1]: I had updated the figure on my machine but had a hard time getting git to recognize that I had changed the file. It should be fixed now.

**Review by PaulTalbot-INL — APPROVED** [review]

**Comment by PaulTalbot-INL** [comment]
Can you convert this from WIP so the automated tests can run? I assume that's what's holding them back.

File: tests/framework/PostProcessors/TSACharacterizer/tests [hunk 9]
```diff
@@ -6,6 +6,7 @@
       type = OrderedCSV
       output = 'Basic/chz.csv'
       windows_gold = 'Basic/windowsChz.csv'
+      mac_gold = 'Basic/windowsChz.csv'
```
Comment (PaulTalbot-INL) [inline tests/framework/PostProcessors/TSACharacterizer/tests:9]: so windows and mac agree but linux differs now?

**Review by PaulTalbot-INL — COMMENTED** [review]
One comment to check.

File: tests/framework/PostProcessors/TSACharacterizer/tests (hunk 9 above)
Comment (j-bryan) [reply tests/framework/PostProcessors/TSACharacterizer/tests:9]: Yeah, seems that the difference is just a phase of one of the Fourier modes being pi on some OSes and -pi on others. Do you want me to change that file name now that that gold file is being used for the Mac test, too?

Comment (PaulTalbot-INL) [reply tests/framework/PostProcessors/TSACharacterizer/tests:9]: Nope, I'm fine with this.

**Review by PaulTalbot-INL — APPROVED** [review]
